File: NutritionSDKBridge.m
//
//  NutritionSDKBridge.h
//  SDKNutritionExample
//
//  Created by Miguel Muñoz on 28/11/23.
//

#import <React/RCTBridgeModule.h>

@interface RCT_EXTERN_MODULE(NutritionSDKBridge, NSObject)

RCT_EXTERN_METHOD(start:(NSString *)user appID:(NSString *)appID password:(NSString *)password resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(isProfileFilled:(RCTPromiseResolveBlock)resolve)

RCT_EXTERN_METHOD(setMemberId:(NSString *)memberId)

RCT_EXTERN_METHOD(setCountry:(Int)country)

RCT_EXTERN_METHOD(startSession:(NSString *)appID password:(NSString *)password)

RCT_EXTERN_METHOD(logout)

RCT_EXTERN_METHOD(showNutritionModule)

RCT_EXTERN_METHOD(setMainColor:(NSString *)color)

RCT_EXTERN_METHOD(setNavigationBarColor:(NSString *)color)

RCT_EXTERN_METHOD(setNavigationTintColor:(NSString *)color)

RCT_EXTERN_METHOD(showNutritionModule)

RCT_EXTERN_METHOD(getProfile:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)

RCT_EXTERN_METHOD(setNavigationTitle:(NSString *)title)

RCT_EXTERN_METHOD(setLanguage:(NSString *)language)

RCT_EXTERN_METHOD(updateProfile:(NSString *)sex height:(nonnull NSNumber *)height weight:(nonnull NSNumber *)weight birthDateString:(NSString *)birthDateString resolver:(RCTPromiseResolveBlock)resolve rejecter:(RCTPromiseRejectBlock)reject)

@end

File: NutritionSDKBridge.swift
//
//  NutritionSDKBridge.swift
//  SDKNutritionExample
//
//  Created by Miguel Muñoz on 28/11/23.
//

import Foundation
import NutritionIASDK
import React

@objc(NutritionSDKBridge)
class NutritionSDKBridge: NSObject {
  
  @objc static func requiresMainQueueSetup() -> Bool {
    return true
  }
  
  @objc func start(_ user: String, appID: String, password: String, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
    NutritionSDK.sharedInstance.start(with: user, appID: appID, password: password)
    resolve(nil)
  }
  
  @objc
  func updateProfile(sex: String?, height: NSNumber?, weight: NSNumber?, birthDateString: String?, resolver resolve: RCTPromiseResolveBlock, rejecter reject: RCTPromiseRejectBlock) {
      let dateFormatter = DateFormatter()
      dateFormatter.dateFormat = "yyyy-MM-dd" // Asegúrate de que este formato coincida con el que envías desde JS

      let birthDate = birthDateString != nil ? dateFormatter.date(from: birthDateString!) : nil
      let gender = sex != nil ? NutritionGender(rawValue: sex!) : nil

      NutritionSDK.sharedInstance.updateProfile(sex: gender, height: height?.intValue, weigth: weight?.intValue, birthDate: birthDate, generateMenu: true, needsWaitingScreens: true)

      resolve(nil)
  }

  @objc
  func showNutritionModule() {
    NutritionSDK.sharedInstance.showNutritionModule()
  }
  
  @objc func isProfileFilled(_ resolve: RCTPromiseResolveBlock) {
    let isFilled = NutritionSDK.sharedInstance.isProfileFilled()
    resolve(isFilled)
  }
  
  @objc func setMemberId(_ memberId: String) {
    NutritionSDK.sharedInstance.setMemberId(memberId)
  }
  
  @objc func setCountry(_ country: Int) {
    // Asumiendo que country se puede convertir a NutritionUserCountry
    let userCountry = NutritionUserCountry(rawValue: country) ?? .uk
    NutritionSDK.sharedInstance.setCountry(userCountry)
  }
  
  @objc func startSession(_ appID: String, password: String) {
    NutritionSDK.sharedInstance.startSession(appID: appID, password: password)
  }
  
  @objc func logout() {
    NutritionSDK.sharedInstance.logout()
  }
  
  @objc func setMainColor(_ color: String) {
    NutritionSDK.sharedInstance.setMainColor(color: color)
  }
  
  @objc
  func setNavigationBarColor(_ color: String) {
    NutritionSDK.sharedInstance.setNavigationBarColor(color: color)
  }

  @objc
  func setNavigationTintColor(_ color: String) {
    NutritionSDK.sharedInstance.setNavigationTintColor(color: color)
  }

  @objc
  func setNavigationTitle(_ title: String) {
    NutritionSDK.sharedInstance.setNavigationTitle(title: title)
  }

  @objc
  func setLanguage(_ language: String) {
    let nutritionLanguage = NutritionLanguage(rawValue: language) ?? .es
    NutritionSDK.sharedInstance.setLanguage(nutritionLanguage)
  }
  
}
